tests/test_repository.py

```diff
@@ -0,0 +1,209 @@
+import os
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Tuple
+
+import pytest
+from prefect.testing.utilities import AsyncMock
+
+import prefect_github
+from prefect_github import GitHubCredentials
+from prefect_github.repository import GitHubRepository
+
+
+class TestGitHubRepository:
+    async def test_subprocess_errors_are_surfaced(self):
+        """Ensure that errors from GitHub are being surfaced to users."""
+        g = GitHubRepository(repository="incorrect-url-scheme")
+        with pytest.raises(
+            OSError, match="fatal: repository 'incorrect-url-scheme' does not exist"
+        ):
+            await g.get_directory()
+
+    async def test_repository_default(self, monkeypatch):
+        """Ensure that default command is 'git clone <repo name>' when given just
+        a repo.
+        """
+
+        class p:
+            returncode = 0
+
+        mock = AsyncMock(return_value=p())
+        monkeypatch.setattr(prefect_github.repository, "run_process", mock)
+        g = GitHubRepository(repository="prefect")
+        await g.get_directory()
+
+        assert mock.await_count == 1
+        assert "git clone prefect" in mock.await_args[0][0]
+
+    async def test_reference_default(self, monkeypatch):
+        """Ensure that default command is 'git clone <repo name> -b <reference> --depth 1'  # noqa: E501
+        when just a repository and reference are given.
+        """
+
+        class p:
+            returncode = 0
+
+        mock = AsyncMock(return_value=p())
+        monkeypatch.setattr(prefect_github.repository, "run_process", mock)
+        g = GitHubRepository(repository="prefect", reference="2.0.0")
+        await g.get_directory()
+
+        assert mock.await_count == 1
+        assert "git clone prefect -b 2.0.0 --depth 1" in mock.await_args[0][0]
+
+    async def test_token_added_correctly_from_credential(self, monkeypatch):
+        """Ensure that the repo url is in the format `https://<oauth-key>@github.com/<username>/<repo>.git`."""  # noqa: E501
+
+        class p:
+            returncode = 0
+
+        mock = AsyncMock(return_value=p())
+        monkeypatch.setattr(prefect_github.repository, "run_process", mock)
+        credential = GitHubCredentials(token="XYZ")
+        g = GitHubRepository(
+            repository="https://github.com/PrefectHQ/prefect.git",
+            credentials=credential,
+        )
+        await g.get_directory()
+        assert mock.await_count == 1
+        assert (
+            "git clone https://XYZ@github.com/PrefectHQ/prefect.git"
+            in mock.await_args[0][0]
+        )
+
+    async def test_ssh_fails_with_credential(self, monkeypatch):
```

Comment:
@ahuang11 I think this should cover the validation, but let me know if you'd like another one.

Reply:
Weird that the coverage is not picking this up
```
======================= 26 passed, 2 warnings in 11.48s ========================
Name                            Stmts   Miss Branch BrPart  Cover   Missing
---------------------------------------------------------------------------
prefect_github/__init__.py          4      0      0      0   100%
prefect_github/credentials.py      13      0      2      0   100%
prefect_github/exceptions.py        2      0      0      0   100%
prefect_github/graphql.py          31      3      8      1    90%   50, 58-59
prefect_github/repository.py      120      0     14      1    99%   [61](https://github.com/PrefectHQ/prefect-github/actions/runs/3176548018/jobs/5175974972#step:5:62)->73
prefect_github/utils.py            30      1     18      2    94%   41->36, 63
---------------------------------------------------------------------------
TOTAL                             200      4     42      4    97%
```